ql/termstructures/yield/nonlinearfittingmethods.cpp

```diff
@@ -32,48 +32,76 @@ namespace QuantLib {
         const ext::shared_ptr<OptimizationMethod>& optimizationMethod,
         const Array& l2,
         const Real minCutoffTime,
-        const Real maxCutoffTime)
+        const Real maxCutoffTime,
+        const Size numCoeffs,
+        const Real fixedKappa)
     : FittedBondDiscountCurve::FittingMethod(
-          constrainAtZero, weights, optimizationMethod, l2, minCutoffTime, maxCutoffTime) {}
+          constrainAtZero, weights, optimizationMethod, l2, minCutoffTime, maxCutoffTime),
+          numCoeffs_(numCoeffs), fixedKappa_(fixedKappa) 
+    {
+        if (numCoeffs_ <= 0)
+            numCoeffs_ = 9;
```

Comment:
Instead of overwriting I'd go for reporting an error, as in
```
QL_REQUIRE(numCoeffs_ > 0, "At least 1 coefficient required");
```
(same for the similar checks below).

Also, I used 0 because that's what your current check implies, but is there a minimum number of parameters we should ask for? Don't we need at least one parameter besides kappa?

Reply:
QL_REQUIRE() is better, I agree: you should only be supplying a number for numCoeffs if you know what you are doing!

As for the minimum number of "coefficients", I was kind of following the terminology from the discountFunction(). The form is df(t) = Coeff1.exp(-Kappa.t) + Coeff2.exp(-2.Kappa.t) + CoeffN.exp(-N.Kappa.t) ... So numCoeffs_ = N ... it is not the number of independent parameters given by the size() method (which could be N+1, N or N-1). You could have a valid, if structureless, curve if numCoeffs = 1 and Kappa is not fixed (ie fixedKappa_ is Null).

We just need to make sure there is one unconstrained variable (be it Kappa or a single coefficient), otherwise the optimization will struggle.

You have four possibilities, which ensure there is at least one unconstrained variable to optimize over:
1. constrainAtZero is true, Kappa is not fixed, so require numCoeffs_ > 0
2. constrainAtZero is true, Kappa is fixed, so require numCoeffs_ > 1
3. constrainAtZero is false, Kappa is not fixed, so require numCoeffs > 0
4. constrainAtZero is false, Kappa is fixed, so require numCoeffs_ > 0

So perhaps
    QL_REQUIRE(numCoeffs_ >( ( constrainAtZero && (fixedKappa_ != Null<Real>() ) ) ? 1 : 0 ), 
                            "At least 1 unconstrained coefficient required");

Reply:
Yes! That is much neater (and obvious now you say it!). Thanks.